xchange-stream-binance/src/main/java/info/bitrich/xchangestream/binance/BinanceStreamingMarketDataService.java

```diff
@@ -300,41 +320,207 @@ private void initRawOrderBookUpdatesSubscription(CurrencyPair currencyPair) {
+
+    private final CompositeDisposable disposables = new CompositeDisposable();
 
-    private OrderbookSubscription(Observable<DepthBinanceWebSocketTransaction> stream) {
-      this.stream = stream;
+    /**
+     * Helps to keep integrity of book snapshot which is initialized and patched on different
+     * threads.
+     */
+    private final Object bookIntegrityMonitor = new Object();
+
+    private OrderBook book;
+    private long bookLastUpdateId;
+
+    private OrderBookSubscription(
+        Observable<DepthBinanceWebSocketTransaction> deltasObservable, CurrencyPair currencyPair) {
+      this.deltasObservable = deltasObservable;
+      this.currencyPair = currencyPair;
     }
 
-    void invalidateSnapshot() {
-      snapshotLastUpdateId.set(0);
+    public Observable<OrderBook> connect() {
+      if (isDisposed())
+        throw new IllegalStateException(
+            "Disposed before, use a new instance to connect next time.");
+
+      disposables.add(asyncInitializeOrderBookSnapshot());
+
+      deltasObservable
+          .doOnNext(
+              delta -> {
+                synchronized (bookIntegrityMonitor) {
+                  if (isBookInitialized()) {
+                    if (!appendDelta(delta)) {
+                      disposables.add(asyncInitializeOrderBookSnapshot());
+                    }
+                  } else {
+                    bufferDelta(delta);
+                  }
+                }
+              })
+          .filter(delta -> isBookInitialized())
+          .map(delta -> getBook())
+          .doFinally(() -> dispose())
+          .subscribe(booksSubject);
+
+      return booksSubject.hide();
     }
 
-    void initSnapshotIfInvalid(CurrencyPair currencyPair) {
-      if (snapshotLastUpdateId.get() != 0) return;
-      try {
-        LOG.info("Fetching initial orderbook snapshot for {} ", currencyPair);
-        onApiCall.run();
-        fallbackOnApiCall.get().run();
-        BinanceOrderbook book = fetchBinanceOrderBook(currencyPair);
-        snapshotLastUpdateId.set(book.lastUpdateId);
-        lastUpdateId.set(book.lastUpdateId);
-        orderBook = BinanceMarketDataService.convertOrderBook(book, currencyPair);
-      } catch (Exception e) {
-        LOG.error("Failed to fetch initial order book for " + currencyPair, e);
-        snapshotLastUpdateId.set(0);
-        lastUpdateId.set(0);
-        orderBook = null;
+    @Override
+    public void dispose() {
+      if (!isDisposed()) {
+        booksSubject.onComplete();
+        disposables.dispose();
+      }
+    }
+
+    private void disposeWithError(Throwable error) {
+      if (!isDisposed()) {
+        booksSubject.onError(error);
+        disposables.dispose();
+      }
+    }
+
+    @Override
+    public boolean isDisposed() {
+      return booksSubject.hasComplete() || booksSubject.hasThrowable();
+    }
+
+    private boolean isBookInitialized() {
+      synchronized (bookIntegrityMonitor) {
```

Comment:
just make the variable volatile, it will use memory barriers, no need to sync for this

Reply:
The case is, `isBookInitialized()` and `getBook()` want the instantiated and fully patched, initial book snapshot. If I would use volatile book reference, the thread observing books, could access instantiated book reference, which is still being patched (in corner cases of cource, but thread safety is abaout eliminating all that corner cases).

Reply:
Maybe there is something you are seing but I can't grasp. I have allowed collaborators to modify my PR, if you like you should be able to propose better synchronization method.

Reply:
I am not following how synchronized would protect against what you are describing. You are simply asking if book is set or not, having a volatile would enforce memory barriers and the thread to read the latest state from memory. It's like an atomic operation this isn't getting updating and setting. It's just an atomic read.

Reply:
That what you are saying, can be true, but I needed to change my thinking, with sync blocks I could assign outdated snapshot to the book reference, then update it. It can be also achieved without sync blocks, just firstly update the snapshot with all the deltas, then assign to the book reference.

Whatever, I'm forking on that :)